### PR TITLE
feat(core): CDB-2535 Implement query filters for database queries

### PR DESCRIPTION
**Pull request by dbcfd**

### CODE REVIEW

**Comment by dbcfd** [comment]
Few more tests I still need to add, but wanted to get some eyes on this to make sure it looks valid.

File: packages/core/src/indexing/__tests__/query-filter-convert.test.ts [hunk 1]
```diff
@@ -0,0 +1,116 @@
+import { jest } from '@jest/globals'
```
Comment (PaulLeCam) [inline packages/core/src/indexing/__tests__/query-filter-convert.test.ts:1]: Can you add tests with actual rows getting returned, checks for range support and other validation as described in https://github.com/ceramicnetwork/js-ceramic/pull/2839#discussion_r1233793048 please?

**Review by PaulLeCam — CHANGES_REQUESTED** [review]
The query building logic is a bit hard to follow, there is duplicated logic such as

```ts
if (first) {
  return query.where(inner)
} else if (combinator && combinator == Combinator.Or) {
  return query.orWhere(inner)
} else {
  return query.andWhere(inner)
}
```

and 

```ts
const isFirst = first
const old = where
where = (bldr) => {
  const b = old(bldr)
  ...
}
```

in multiple places that look like they could be refactored and simplified?

File: packages/common/src/index-api.ts [hunk 2]
```diff
@@ -75,10 +75,17 @@ export type RangeValueFilter<T extends string | number> =
...
   RangeValueFilter<T>
 
+type StringCommonValueFilter = ScalarValueFilter<string>
+type NumberCommonValueFilter = ScalarValueFilter<number>
+
 /**
  * Any supported field value filter on an object
  */
-export type AnyValueFilter = BooleanValueFilter | EnumValueFilter | ScalarValueFilter
+export type AnyValueFilter =
```
Comment (PaulLeCam) [inline packages/common/src/index-api.ts:84]: This change isn't needed, `ScalarValueFilter` already works for both `string` and `number`

**Review by PaulLeCam — DISMISSED** [review]

File: packages/common/src/index-api.ts [hunk 3]
```diff
@@ -28,19 +28,19 @@ export type Pagination = ForwardPagination | BackwardPagination
  * Boolean field value filter
  */
 export type BooleanValueFilter =
-  | { isNull: boolean }
-  | { equalTo: boolean }
-  | { notEqualTo: boolean }
+  | { type: 'bool'; op: 'null'; value: boolean }
```
[the remark this answers is not in the document]
Comment (dbcfd) [reply packages/common/src/index-api.ts:31]: Since js-ceramic is a different system than js-composedb, this is the proposed public API for js-ceramic, since it does not utilized GraphQL. This is due to the separation of concerns that exists between the two systems

- js-ceramic: Types that provide information to construct queries for filtering, that can be easily represented in a REST request
- js-composedb: Graphql types that represent a query

The API of js-composedb will use types, that are represented as GraphQL for its inputs, while js-ceramic will use these types as its inputs. It is the responsibility of js-composedb to convert its inputs to inputs that js-ceramic requires.

It is possible there exists a type that requires no translation between the two systems, but the current implementation in js-ceramic uses the types in this PR, as they contain enough information to implement the necessary functionality, while being uniform enough to provide an understandable API for js-ceramic.

Please feel free to propose an alternate type for js-ceramic that meets its concerns.

[the remark this answers is not in the document]
Comment (PaulLeCam) [reply packages/common/src/index-api.ts:31]: I think this is going backwards. What value does changing the types here bring, when we already discussed them in a previous PR (https://github.com/ceramicnetwork/js-ceramic/pull/2839), which was an explicit requirement to enable other work (https://github.com/ceramicstudio/js-composedb/pull/135)?
How does having different data shapes for filters in ComposeDB and Ceramic APIs serve our users in any way?

[the remark this answers is not in the document]
Comment (dbcfd) [reply packages/common/src/index-api.ts:31]: The value here is that this is what is needed for js-ceramic to implement the required functionality. The previous PR said these are the filters that js-composedb would like to have, this PR builds on that and says this is the shape of those filters so the functionality can be implemented.

As far as serving users, this is the format for the Ceramic API that achieves the functionality users want, in a way that corresponds to its exposed interface. ComposeDB is not a concern of Ceramic, since the functionality that ComposeDB is implementing can be achieved with this API.

Again, if there is a format for these filters that exists to achieve this implementation without putting additional burden on users of the Ceramic API, please propose it.

[the remark this answers is not in the document]
Comment (stbrody) [reply packages/common/src/index-api.ts:31]: @dbcfd can you explain why these types are better than the existing types from the perspective of implementing the ceramic-side functionality?  They seem to be directly isomorphic to the existing types.  I agree that that there is a separation of concerns between Ceramic and Composedb, but if there's already a 1-1 mapping between the two types, why not just stick with the types that composedb uses for simplicity sake?

Also, I don't think that the logical "composedb" boundary ends at the beginning of the js-ceramic codebase, there is already a lot of composedb-specific code in js-ceramic.  I more and more already think of js-ceramic as the "composedb" server-side node.  But that's a bigger conversation for a different time.

[the remark this answers is not in the document]
Comment (dbcfd) [reply packages/common/src/index-api.ts:31]: - Ease of use. For people using the api, this is a consistent interface for each query they input. For example, it's easier to set `{op: '=', value: 1}` than to remember `{equal: 1}` or is it `{eq: 1}`? And then if I need greater than, it's `{op='>', value: 1}`.
- Strictness of type: Without the type hint on the union, we may have to resort to any casting 
- Ease of implementation. Since this is a union field, we can build queries using the type, op, and value, rather than checking for individual fields to see if they are present. Although typescript might not allow that either.

This seemed like a good compromise between a stronger interface like codecs where we have to type each of the queries, and what was originally proposed as needed by composedb.

[the remark this answers is not in the document]
Comment (stbrody) [reply packages/common/src/index-api.ts:31]: > Ease of use. For people using the api, this is a consistent interface for each query they input. For example, it's easier to set {op: '=', value: 1} than to remember {equal: 1} or is it {eq: 1}? And then if I need greater than, it's {op: '>', value: 1}.

That seems pretty subjective.  This syntax is quite a bit more verbose, so I think some would say its less easy to use because of that.

> * Strictness of type: Without the type hint on the union, we may have to resort to any casting

Can't speak to that

> * Ease of implementation. Since this is a union field, we can build queries using the type, op, and value, rather than checking for individual fields to see if they are present. Although typescript might not allow that either.

This one I definitely *do* buy.  This format actually reads to me like an internal representation specifically designed for ease of implementation.  I could even see a world where the public API is the existing types used by graphql but then we translate that to a representation like this internally.


Overall, I ultimately don't feel *super* strongly about this one way or the other.  I do lean towards keeping the types the same between ComposeDB and Ceramic, as I think doing so may make things easier as we move to combine the codebases and move more composedb functionality server-side over time.  But if you and Paul can reach a consensus then I'm ultimately happy either way.

File: packages/core/src/indexing/insertion-order.ts [hunk 4]
```diff
@@ -128,9 +129,11 @@ export class InsertionOrder {
     pagination: ForwardPaginationQuery
   ): Knex.QueryBuilder<unknown, Array<Selected>> {
     const tableName = asTableName(query.model)
+    const converted = convertQueryFilter(query.queryFilters)
     let base = this.dbConnection
       .from(tableName)
-      .select('stream_id', 'last_anchored_at', 'created_at')
+      .select(converted.select.concat(['stream_id', 'last_anchored_at', 'created_at']))
```
[the remark this answers is not in the document]
Comment (dbcfd) [reply packages/core/src/indexing/insertion-order.ts]: Yeah, originally I was only returning fields of interest, but realized that what we query on may not be what the user wants to see. Changing this to return just stream_content

**Comment by bigs** [comment]
Hey y'all, seems like we're at a bit of an impasse with regards to our API design. I'm looking forward to our time together tomorrow morning so we can talk things through. We'll have to make a decision as to whether this API remains closer to the implementor or to the user—there are advantages to each. I'd like to see us come to a conclusion that we can all live with, rather than one that feels perfect. Also—we'll talk about it on our call tomorrow—there will be an opportunity to give this codebase another nudge in the near future should we want to change our approach.

File: packages/common/src/index-api.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (PaulLeCam) [reply packages/common/src/index-api.ts:31]: As Spencer wrote, these APIs are designed for ComposeDB, it's not like there is another consumer for these we need to care about, and the different type structures are semantically the same.

I don't mind whatever internal representation is used, but it seems unnecessary to me. There's already a `convertQueryFilter()` that converts the API representation to the SQL representation so it's not like the suggested changes represent the canonical form of the filters for the DB anyways, it's just another abstraction.
I'd much rather have `GraphQL inputs as API structure representation` -> `SQL conversion` than `GraphQL inputs format` -> `intermediary API representation` -> `SQL conversion` where the intermediary representation doesn't bring any additional benefit, it's much easier to follow the logic, test and maintain.

File: packages/core/src/indexing/database-index-api.ts [hunk 5]
```diff
@@ -43,7 +43,7 @@ export interface IndexModelArgs {
 type IndexedData<DateType> = {
   stream_id: string
   controller_did: string
-  stream_content: any
+  stream_content: Record<string, any> | string
```
Comment (willex) [inline packages/core/src/indexing/database-index-api.ts:46]: Could this potentially be just a `Record`? Can be addressed in a different PR though.

**Review by willex — APPROVED** [review]
@dbcfd Looks good on my end. Seems the issues from the PR review discussion have been addressed.

File: packages/core/src/indexing/insertion-order.ts [hunk 6]
```diff
@@ -128,27 +133,13 @@ export class InsertionOrder {
...
-    }
-    if (query.filter) {
-      for (const [key, value] of Object.entries(query.filter)) {
-        const filterObj = {}
-        filterObj[addColumnPrefix(key)] = value
-        base = base.andWhere(filterObj)
-      }
-    }
+    const queryFunc = this.query(query, false)
+    let base = queryFunc(this.dbConnection.from(tableName)).limit(pagination.first + 1)
```
Comment (stbrody) [inline packages/core/src/indexing/insertion-order.ts:137]: nice cleanup

File: packages/core/src/indexing/query-filter-converter.ts [hunk 7]
```diff
@@ -0,0 +1,224 @@
...
+
+function nextState<Old, New>(state: ConversionState<Old>, next: New): ConversionState<New> {
+  return {
+    filter: next,
+    negated: state.negated,
+    combinator: state.combinator,
+  }
+}
+
+function handleQuery(
```
Comment (stbrody) [inline packages/core/src/indexing/query-filter-converter.ts:28]: would be really nice to have some comments on this explaining the different args and how they are meant to be used

File: packages/core/src/indexing/query-filter-converter.ts [hunk 8]
```diff
@@ -0,0 +1,224 @@
...
+    }
+  }
+  if (negated) {
+    return query.whereNot(opFunc)
+  } else {
+    return query.where(opFunc)
+  }
+}
+
+function handleIn<T extends number | string>(
```
Comment (stbrody) [inline packages/core/src/indexing/query-filter-converter.ts:55]: ditto re: comment for the args

File: packages/core/src/indexing/query-filter-parser.ts [hunk 9]
```diff
@@ -0,0 +1,159 @@
...
+  } else {
+    return 'string'
+  }
+}
+export function parseObjectFilter(filter: ApiObjectFilter): ObjectFilter {
+  const entries: Record<string, AllValueFilter> = {}
+  const keys = Object.keys(filter)
+  for (const key of keys) {
+    const childFilter = filter[key]
+    if ('isNull' in childFilter) {
```
Comment (stbrody) [inline packages/core/src/indexing/query-filter-parser.ts:59]: this doesn't actually do what we discussed on the call of using type assertions to do these checks so we can leverage the type system more natively.  That said I don't personally feel strongly about this so I'm fine with it if it works for @PaulLeCam and @bigs

**Review by stbrody — COMMENTED** [review]
LGTM, main request is for more comments throughout, this code is pretty complex and not the most intuitive
